### DISCUSSION TITLE
Obtaining Response Content Length & IP+Port within middleware?

### DISCUSSION THREAD

**Discussion by rickmartensnl** [actix/actix-web#2983] [Q&A] [answered]
I am currently working on developing a middleware using wrap (and not wrap_fn) to log data in the JSON format specified here: https://cloud.google.com/logging/docs/agent/logging/configuration#special-fields. While I have made progress in resolving various variables, I have a couple of questions:

1. Is it possible to retrieve the length of the response content from within the middleware?
2. Is it possible to retrieve the bound IP address and port of the server from within the middleware?

Thank you for your assistance in advance.

**Comment by robjtede** [comment, marked as answer]
1. Yes, see the Logger middleware for an example. In the general case, you'll need to read the entire payload stream to know for sure; you can't rely on there being a Content-Length header or even the Payload size_hint for this.
2. Yes, through [HttpRequest::app_config()](https://docs.rs/actix-web/latest/actix_web/struct.HttpRequest.html#method.app_config) method (admittedly with lack-luster docs).

**Reply by rickmartensnl** [reply]
Do you have a link to the specific line? I can't really find the logic of the calculation of the logger middleware.
Thanks!

**Reply by robjtede** [reply]
Sure! This section is the body stream wrapper that does the byte counting: https://docs.rs/actix-web/4.3.0/src/actix_web/middleware/logger.rs.html#401-453

**Reply by rickmartensnl** [reply]
Still can't figure it out, can you help me out? This is my logger:
https://gist.github.com/rickmartensnl/1024adc3a922cb7146b5f15d9ce3bb5a

**Reply by robjtede** [reply]
One of the key points here is that all the state you want to log must be part of the stream wrapper (again, see `StreamLog`) and the log occurs when it is dropped. There's no other way to get the response size without buffering the entire response body.
